### PR TITLE
[events executor] - Fix Behavior with Timer Cancel

### PR DESCRIPTION
**Pull request by mwcondino**
## Background
It appears that with the experimental events executor, some usages of the rclcpp API cause broken behavior. Specifically, if multiple timers are configured in a node, and during runtime one of the timers is cancelled via the `TimerBase::cancel` method, this will result in all other timers failing to execute.

A minimal working example of the bug is found here --
[example_package.tar.gz](https://github.com/ros2/rclcpp/files/13494070/example_package.tar.gz)

To reproduce, build the package using `colcon build`, then run `USE_EVENTS_EXECUTOR=1 ros2 run example_package timer_node`. It is observable that both timers get cancelled after the first timer cancels itself:
```
root@753c2035d766:/app/src# USE_EVENTS_EXECUTOR=1 ros2 run example_package timer_node
[INFO] [1701211973.238745755] [timer_node]: Using EventsExecutor
[INFO] [1701211974.239147118] [timer_node]: Timer 1!
[INFO] [1701211974.239268496] [timer_node]: Timer 2!
[INFO] [1701211975.239086741] [timer_node]: Timer 1!
[INFO] [1701211975.239251090] [timer_node]: Timer 2!
[INFO] [1701211976.239241288] [timer_node]: Timer 1!
[INFO] [1701211976.239417539] [timer_node]: Timer 2!
[INFO] [1701211977.239107783] [timer_node]: Timer 1!
[INFO] [1701211977.239260069] [timer_node]: Timer 2!
[INFO] [1701211978.239079416] [timer_node]: Timer 1!
[INFO] [1701211978.239232915] [timer_node]: Timer 2!
[INFO] [1701211979.239045079] [timer_node]: Timer 1!
[INFO] [1701211979.239198617] [timer_node]: Timer cancelling itself!
[INFO] [1701211979.239225057] [timer_node]: Timer 2!
```

Myself and @gusbrigantino  have tested this on latest `rolling` as well as `iron`.

## Solution
The root cause is that there’s an edgecase within the `TimersManager` class which causes the thread to hang indefinitely if the ‘head’ timer in an internal data structure gets cancelled. As a result, the method will block until timers are reset (the `timers_updated_` flag is set, which only occurs if the timers are reset. The underlying issue is that the `TimersManager` class is not properly checking for if a timer to be executed has actually been cancelled.

The implemented fix adds logic to the comparison used for the heap such that cancelled timers are always at the bottom of the heap. Furthermore,  `get_head_timeout_unsafe` is modified to return whether or not the current timer is invalid. This bool is then used to break out of the loop in `run_timers`.

### CODE REVIEW

**Comment by mwcondino** [comment]
Looks like the build for `rolling` may be unhappy? I'm seeing a similar build failure for the latest merge into `rolling` - [link](https://github.com/ros2/rclcpp/pull/2374/commits)

**Comment by mjcarroll** [comment]
> Looks like the build for rolling may be unhappy? I'm seeing a similar build failure for the latest merge into rolling - [link](https://github.com/ros2/rclcpp/pull/2374/commits)

You can probably disregard `Rpr` jobs right now, we will run full CI on it.

**Comment by mjcarroll** [comment]
Is there any way that you can add your reproducible example as a test?

**Comment by mwcondino** [comment]
> Is there any way that you can add your reproducible example as a test?

Sure, I can try to get it in the form of a unit test.

File: rclcpp/src/rclcpp/experimental/timers_manager.cpp
Comment (alsora) [inline rclcpp/src/rclcpp/experimental/timers_manager.cpp:195]: do we need an extra variable? what does `time_until_trigger` returns when the timer is cancelled?

File: rclcpp/include/rclcpp/experimental/timers_manager.hpp
Comment (alsora) [inline rclcpp/include/rclcpp/experimental/timers_manager.hpp:499]: I see, this seems a bug I introduced when porting the code to rclcpp.
in our internal version, we added a new function `next_call_time` which takes into account cancellations
https://github.com/irobot-ros/rclcpp/blob/galactic-events-executor/rclcpp/src/rclcpp/timer.cpp#L145-L164

File: rclcpp/src/rclcpp/experimental/timers_manager.cpp
Comment (mwcondino) [reply rclcpp/src/rclcpp/experimental/timers_manager.cpp:195]: Yes, I found it was necessary. `time_until_trigger` actually returns `std::chrono::nanoseconds::max()` when the timer is cancelled, which I think is how the bug manifests -- that then forces the call to `timers_cv_.wait(...)`

File: rclcpp/include/rclcpp/experimental/timers_manager.hpp
Comment (mwcondino) [reply rclcpp/include/rclcpp/experimental/timers_manager.hpp:499]: I think even if `next_call_time` was used, the bug would still manifest -- that would cause a situation where we wait on `timers_cv_.wait(lock, [this]() {return timers_updated_;});`, which won't return until a timer is manually reset.

An alternative fix that we considered was modifiying `rcl` and the timer object to support a cancellation callback, which would then be used similarly to how `set_on_reset_callback` is used in `TimersManager`. If we could trigger a callback upon the timer being reset, then I think that would allow us to break the condition and then re-order the heap, likely fixing the issue.

Comment (alsora) [reply rclcpp/include/rclcpp/experimental/timers_manager.hpp:499]: Ok, good point.
I think we should go ahead with the current fix for now. Adding a new callback for timers cancelled is doable, but it's a larger effort.
Any thoughts @mauropasse ?

**Comment by mwcondino** [comment]
OK I ended up adding two tests:
1. `check_one_timer_cancel_doesnt_affect_other_timers` in `test_timers_manager.cpp` validates the behavior in the timers_manager object. Interestingly enough, this test **does not fail** on latest `iron`, which I did not expect given the example package's behavior
2. `testOneTimerCancelledWithExecutorSpin` in `test_executors.cpp` validates the behavior when using a `rclcpp::Node` that adds and then cancels it's own timers. This example uses `rclcpp::spin`, and **does** reproduce the behavior. On this branch, this test should be passing.

I can't figure out why there would be a difference in behavior between the two tests - as far as I can tell, when `spin` is called on the `EventsExecutor`, it makes a call to `start` it's own `timers_manager` object, so I would think that the behavior should be the same as when the `timers_manager` object is manually started and then the same sequence of waiting + cancelling is followed.

File: rclcpp/include/rclcpp/experimental/timers_manager.hpp
Comment (gusbrigantino) [reply rclcpp/include/rclcpp/experimental/timers_manager.hpp]: We ended up reverting this diff here in the `timer_greater` function because we discovered it was no longer necessary for the bug fix

File: rclcpp/src/rclcpp/experimental/timers_manager.cpp
Comment (alsora) [reply rclcpp/src/rclcpp/experimental/timers_manager.cpp:195]: It looks like the `get_head_timeout_unsafe` is essentially trying to indicate 3 different scenarios:
 - it should indicate when the heap is empty, and it currently does that by returning `std::chrono::nanoseconds::max()`
 - it should indicate when the front timer in the heap is cancelled, and it currently does that by setting the `head_was_cancelled` bool variable
 - if the front timer is good, then it should indicate the time until trigger

I think that this behavior is not very intuitive.

Moreover, what happens if there's only 1 timer which is cancelled? Will the heap keep polling it?

File: rclcpp/src/rclcpp/experimental/timers_manager.cpp
Comment (alsora) [inline rclcpp/src/rclcpp/experimental/timers_manager.cpp:256]: What happens if there's only 1 timer which is cancelled? Will the heap keep entering here at every iteration, without ever sleeping?

File: rclcpp/src/rclcpp/experimental/timers_manager.cpp
Comment (mwcondino) [reply rclcpp/src/rclcpp/experimental/timers_manager.cpp:195]: Great point! Refactored to use `std::optional` as the return value.

File: rclcpp/src/rclcpp/experimental/timers_manager.cpp
Comment (alsora) [reply rclcpp/src/rclcpp/experimental/timers_manager.cpp]: I don't think that the fix works.
What if there are more than 1 timers but they are all cancelled?

I think that the correct approach would likely require to:
a) remove cancelled timers from the heap
b) have a notification to put them back in when they are re-activated

Comment (mwcondino) [reply rclcpp/src/rclcpp/experimental/timers_manager.cpp]: @alsora sorry for the delay here - had a busy couple of weeks.

@gusbrigantino and I refactored the code, and we came up with an implementation of `run_timers` that does not require storing cancelled timers and removing them from the heap. Rather, the approach is to try to re-heap if the head timer was cancelled, then re-organize some logic from that point on. If there are non-cancelled timers in the heap, then the re-heap will force them to be at the front and thus head is now valid and the logic remains the same. However, if all timers in the heap are cancelled, the behavior now matches the previous behavior with no timers in the heap -- we wait until `timers_updated` is `true`.

We added a few more test cases to validate a couple of different scenarios.

Please let us know if that makes sense and works for you!

File: rclcpp/include/rclcpp/experimental/timers_manager.hpp
Comment (fujitatomoya) [inline rclcpp/include/rclcpp/experimental/timers_manager.hpp:283]: this method is not used currently, so should be removed from this PR?

Comment (mwcondino) [reply rclcpp/include/rclcpp/experimental/timers_manager.hpp]: Good call, removed

**Comment by mwcondino** [comment]
@alsora @mjcarroll could someone please review whenever convenient? Thanks!

**Comment by fujitatomoya** [comment]
@mwcondino FYI, besides unresolved comments, there are several failures you can find https://build.ros2.org/job/Rpr__rclcpp__ubuntu_jammy_amd64/1272/testReport/

**Comment by mwcondino** [comment]
> @mwcondino FYI, besides unresolved comments, there are several failures you can find https://build.ros2.org/job/Rpr__rclcpp__ubuntu_jammy_amd64/1272/testReport/

Thank

> @mwcondino FYI, besides unresolved comments, there are several failures you can find https://build.ros2.org/job/Rpr__rclcpp__ubuntu_jammy_amd64/1272/testReport/

Thank you, appreciate the heads up! I am new to contributing to ROS repos, so I was unsure of which failures were important (I think the main build is still messed up, but that seems unrelated to this PR).

Linting should be fixed now.

File: rclcpp/src/rclcpp/experimental/timers_manager.cpp
Comment (alsora) [inline rclcpp/src/rclcpp/experimental/timers_manager.cpp:256]: The queue is sorted according to the `return a->time_until_trigger() > b->time_until_trigger();` comparison.
What does `time_until_trigger` returns for a cancelled timer?

**Comment by alsora** [comment]
CI run 
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20074/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14523/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20793/)

File: rclcpp/src/rclcpp/experimental/timers_manager.cpp
Comment (mwcondino) [reply rclcpp/src/rclcpp/experimental/timers_manager.cpp:256]: Upon a cancelled timer, `time_until_trigger` returns `std::chrono::nanoseconds::max()`, so resorting should force cancelled timers to be at the bottom of the heap. In a previous iteration of the PR, we had updated the comparison function for the heap to explicitly check for canceled timers and ensure they were on the bottom of the heap, but that was removed since it's obsolete if `std::chrono::nanoseconds::max()` is always returned for a cancelled timer.

**Review by alsora — APPROVED** [review]
The changes look good to me.
I would like @mauropasse to take a look as well before merging it.

**Comment by mauropasse** [comment]
The fix and unit tests look good to me!

**Comment by fujitatomoya** [comment]
windows are failing https://ci.ros2.org/job/ci_windows/20793/, i think that is because of timer accuracy. (e.g https://ci.ros2.org/job/ci_windows/20793/testReport/rclcpp/TestTimerCancelBehavior_SingleThreadedExecutor/testHeadTimerCancelThenResetBehavior/ timer does not get called as expected.)

**Comment by mwcondino** [comment]
> windows are failing https://ci.ros2.org/job/ci_windows/20793/, i think that is because of timer accuracy. (e.g https://ci.ros2.org/job/ci_windows/20793/testReport/rclcpp/TestTimerCancelBehavior_SingleThreadedExecutor/testHeadTimerCancelThenResetBehavior/ timer does not get called as expected.)

Thanks for the callout -  I relaxed the timing constraints, hopefully that will fix things

**Comment by clalancette** [comment]
> Thanks for the callout - I relaxed the timing constraints, hopefully that will fix things

They should be a *lot* more relaxed than that, otherwise we are going to end up with flaky tests.

In general, on Windows, you should expect things to take 4x as long as on Linux.  We shouldn't even consider having timing tests with constraints of less than 200ms or so.

**Comment by alsora** [comment]
New build after changes in unit-tests:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20089/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14536/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20809/)


It's probably for a separate discussion, but I really feel unease with having so many unit-tests using wall-time clock.
IMO unit-tests should always use a simulated time, to be robust to latency that happens in the server.

**Comment by clalancette** [comment]
> It's probably for a separate discussion, but I really feel unease with having so many unit-tests using wall-time clock.

I agree to an extent.  We probably should have more tests that use some kind of simulated time.  The main reason we don't is that those are harder tests to write.

But I also think that we are always going to have to have tests that use system time.  They show problems that a simulated time won't, and they are what most of our users are using.

**Comment by mwcondino** [comment]
Updated the test to use a simulated clock. For now, that is at .25 realtime, but we can update as necessary if the tests are still flaky and require a slower rate.

**Comment by mwcondino** [comment]
@alsora could you please trigger a full CI run whenever convenient?

**Comment by alsora** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/20129/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/14568/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20841/)

**Comment by mwcondino** [comment]
Looks like there's still some windows failures, but they appear unrelated to this PR (unless I'm missing something):
[projectroot.test_n_nodes__rmw_fastrtps_cpp](https://ci.ros2.org/job/ci_windows/20841/testReport/junit/(root)/projectroot/test_n_nodes__rmw_fastrtps_cpp/)
[test_rclcpp.TestNNodes.test_10_nodes](https://ci.ros2.org/job/ci_windows/20841/testReport/junit/test_rclcpp/TestNNodes/test_10_nodes/)
[test_rclcpp.TestNNodesAfterShutdown.test_10_nodes](https://ci.ros2.org/job/ci_windows/20841/testReport/junit/test_rclcpp/TestNNodesAfterShutdown/test_10_nodes/)

**Comment by clalancette** [comment]
> Looks like there's still some windows failures, but they appear unrelated to this PR (unless I'm missing something):

While I generally agree with you (tests will only be using the events executor if they are explicitly configured to do so), it is kind of weird that these are failing.  We don't see those failures in the nightlies: https://ci.ros2.org/view/nightly/job/nightly_win_rel/ .

I think we should re-run Windows, and if it comes back green this time we can go ahead with it.

**Comment by mwcondino** [comment]
> I think we should re-run Windows, and if it comes back green this time we can go ahead with it.

Sounds good - is it possible for me to kick off the jobs? I feel bad bugging folks to start the CI, but I can't see any buttons on https://ci.ros2.org/job/ci_windows/ to start things, and I'm wondering if I'm missing anything (admittedly I have not used Jenkins in the past so this could definiely be a noob question)

**Comment by alsora** [comment]
Currently only maintainers can trigger a job, and don't hesitate to ping if we forget to do it!

Windowns-only job:
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/20889/)

**Comment by alsora** [comment]
Windows CI is green, indicating that the previous failure is unrelated.
Merging this PR

**Pull request merged**
